### PR TITLE
MAINT: `_lib`: co-vendor array-api-extra and array-api-compat

### PR DESCRIPTION
**Pull request by crusaderky**
array-api-extra has introduced a dependency to array-api-compat (https://github.com/data-apis/array-api-extra/pull/47).

### CODE REVIEW

**Comment by crusaderky** [comment]
> workflows awaiting approval

@lucascolley could you unlock CI please?

**Comment by lucascolley** [comment]
perhaps the hook file needs to be added as a Python source in `_lib/meson.build`?

File: scipy/_lib/_array_api_compat_vendor.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# DO NOT RENAME THIS FILE
+# This is a hook for array_api_extra/src/array_api_extra/_lib/_compat.py
+# to override functions of array_api_compat.
+
+from __future__ import annotations
```
Comment (lucascolley) [inline scipy/_lib/_array_api_compat_vendor.py:5]: is this line required? https://github.com/scipy/scipy/pull/21981 purged this from SciPy, and https://github.com/data-apis/array-api-extra/pull/61 did the same for array-api-extra. Unless it is required for an interaction with array-api-compat?

**Review by lucascolley — APPROVED** [review]
LGTM if CI is happy!

**Comment by crusaderky** [comment]
> FAILED scipy/differentiate/tests/test_differentiate.py::TestHessian::test_example[shape0-numpy] - AttributeError: 'numpy.ndarray' object has no attribute 'device'

I'm baffled by this. How can it be green in main (numpy 1+array-api-compat 1.9.1)?!?

**Comment by lucascolley** [comment]
> DNM: the two submodules need to be repointed to the next tagged releases of array-api-compat and array-api-extra

I think this isn't so important - we can merge and then bump again when the releases are out

**Comment by lucascolley** [comment]
smoke-docs failure looks real

**Comment by lucascolley** [comment]
if you want to submit that `__init__` patch I can cut a release

**Comment by lucascolley** [comment]
let's put this in to unblock gh-21783 - we can bump array-api-compat again when it cuts a release.

Thanks @crusaderky !

**Pull request merged**